### PR TITLE
Packages: Various improvements for wp.com or self-contained consumers

### PR DESCRIPTION
**Pull request by kraftbj**
Various fixes for issues found while adding packages to WP.com, most would apply to any non-Jetpack use of packages.

#### Changes proposed in this Pull Request:
* Move functionality into packages out of the Jetpack class:
* Moved `Jetpack::is_staging_site` into the `Status` package and updates uses to point to that.
* Adds filters, `jetpack_get_file_for_environment`, in the `Assets` package to allow filtering by WP.com to locations specific to that platform.
* Updates tests in the `Assets` package to check the filter. Uses the Brain\Monkey system to mock WP functions without a lot of added weight. This system is PHP 5.6 friendly, which is why I use it over things like WP_Mock.
* Adds translator comment in the `Connections` class to pass WP.com linting.
* Updates the tests in the Status class for the `is_staging_site` function plus convert to Brain\Monkey.
* In the `Sync` package, add a `WPCOM` check as a hard block to running on the WP.com platform.
* Updates the `Sync` package to only include whitelisted constants for Jetpack SSO if within Jetpack (by the presence of the SSO file being able to be included).
* Updated the plugin's tests to reflect change to `is_staging_site`.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* n/a

#### Testing instructions:
* 

#### Proposed changelog entry for your changes:
*

### CODE REVIEW

**Comment by kraftbj** [comment]
These changes are live via r200590-wpcom and would need to be merged before `composer update` there.

**Comment by jeherve** [comment]
I'm seeing a few notices when I check out that branch:

```
Notice: Automattic\Jetpack\Sync\Actions was called incorrectly. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Actions Please see Debugging in WordPress for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4903

Notice: Automattic\Jetpack\Sync\Settings was called incorrectly. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Settings Please see Debugging in WordPress for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4903

Notice: Automattic\Jetpack\Sync\Defaults was called incorrectly. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Defaults Please see Debugging in WordPress for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4903

Notice: Automattic\Jetpack\Status was called incorrectly. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Status Please see Debugging in WordPress for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4903

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php:4903) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 5946

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php:4903) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-admin/includes/misc.php on line 1252

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php:4903) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-admin/admin-header.php on line 9
```

**Comment by kraftbj** [comment]
I'll dig into those errors. Likely @zinigor's work with init will help here, but I'll go through it in more detail.

With 0f65b5e, @jeherve, do we have a convention on filters in packages yet, namely the docblock so it renders correctly on developer.jetpack.com? Can we act like packages are modules for the sake of the parser or do you think we need to handle them differently?

0f65b5e would be in support of resolving what let to the issues with r200737-wpcom, namely that WP.com has a special version of the function that we won't be able to run as we move to packages, so wp.com could filter it instead.

cc: @westi to close loop after your assistance this weekend.

**Review by jeherve — COMMENTED** [review]
> do we have a convention on filters in packages yet, namely the docblock so it renders correctly on developer.jetpack.com? Can we act like packages are modules for the sake of the parser or do you think we need to handle them differently?

That's a good question! I'd be okay with mixing the two, but I think having to prefix with `@module` may be confusing for new contributors. Could we use `@package xx`, so the format is familiar for everyone? We can update the parser to take that into account and create a new taxonomy to handle it.

**Comment by kraftbj** [comment]
I'm just going to ignore the Status class for now. https://github.com/Automattic/jetpack/pull/14168 would remove the Sync/Main::init() which is what is causing Status to be called early.

**Comment by jeherve** [comment]
I am now getting other notices :\

```
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Sync\Settings was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Settings Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Sync\Defaults was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Defaults Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Status was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Status Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Sync\Actions was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Actions Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Sync\Settings was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Settings Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Sync\Defaults was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Sync\Defaults Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
[17-Dec-2019 11:56:57 UTC] PHP Notice:  Automattic\Jetpack\Status was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\Jetpack\Status Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-retry/phpcs-changed.) in /srv/users/user564e54a8/apps/user564e54a8/public/wp-includes/functions.php on line 4970
```

This happens even when all other plugins are deactivated.

**Comment by kraftbj** [comment]
Let's hold off until after 14168 has landed. I didn't get these immediately after applying the fix but in either case, I feel like we're going to go down a rabbit hole that Igor is already at the bottom of. At least gain the wins he's already working on first.

**Comment by kraftbj** [comment]
After leaving things running, I see those errors too.

File: _inc/lib/admin-pages/class.jetpack-react-page.php
Comment (gravityrail) [inline _inc/lib/admin-pages/class.jetpack-react-page.php]: Now that you're creating an instance here, shouldn't the line 239 be something like:

```php
'isActive' => $status->is_development_mode(),
```

File: modules/sso.php
Comment (gravityrail) [inline modules/sso.php]: I don't understand why we're using two invocation patterns in this PR: `( new Status() )->some_function()` vs `$status->some_function()`.

Am I missing something about where/how they should be used?

**Review by gravityrail — DISMISSED** [review]
Epic patch. Would be nice to see some of this either documented more extensively in the description or broken out into individual PRs.

File: modules/sso.php
Comment (kraftbj) [reply modules/sso.php]: When there's only one usage of the package class in a function, I've stuck with `( new Status()->some_function()`. When the same class is used multiple times, I've set it with a variable.

e.g. _inc/lib/class.core-rest-api-endpoints.php L981 and L984 in the same function call two separate status checks. Where modules/sso.php L151 is the only check within that function.

I don't know that it matters really since Status has no constructor, but in general, I would think it is more performant not to construct multiple class instances within the same function.

File: _inc/lib/admin-pages/class.jetpack-react-page.php
Comment (kraftbj) [reply _inc/lib/admin-pages/class.jetpack-react-page.php]: Good catch. Removing the instance in a948dac66 since Status is only being used once.

**Comment by kraftbj** [comment]
@gravityrail:

> Would be nice to see some of this either documented more extensively in the description or broken out into individual PRs.

That's valid. I wrapped them in a single PR since they all need to land for the packages to run on WP.com and majority of the PR is for the `is_staging_site` changes. I've updated the description of the PR to better explain what is changed.

**Comment by kraftbj** [comment]
Rebased and removed Status from the autoload violation list pending testing after 14186.

**Comment by kraftbj** [comment]
After 14186, no notices in the logs and things are working as expected. I'd appreciate extra eyes though (@zinigor want to give it a spin ;-)? )

**Review by jeherve — CHANGES_REQUESTED** [review]
Looks like you'll need to update the use of `is_staging_site` in `_inc/lib/admin-pages/class.jetpack-react-page.php` to avoid notices like this when loading Jetpack > Dashboard:

```
[13-Jan-2020 10:48:12 UTC] PHP Notice:  Jetpack::is_staging_site is <strong>deprecated</strong> since version jetpack-8.1! Use /Automattic/Jetpack/Status->is_staging_site instead. in /srv/users/user97b17710/apps/user97b17710/public/wp-includes/functions.php on line 4648
```

**Comment by kraftbj** [comment]
I rebased that one out--thanks for catching it @jeherve.

**Review by jeherve — APPROVED** [review]
This seems to test well for me now. 👍

**Pull request merged**
